Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp

```diff
@@ -750,12 +750,82 @@ class OMGIRGenerator {
         return WTFMove(m_exceptionHandlers);
     }
 
+    struct AbstractHeap {
+        enum Kind : uint8_t {
+            Globals,
+            WasmMemory,
+            CallSiteIndex,
+            WasmTables,
+            WasmImports,
+            InstanceFields,
+            TableFields,
+            FuncRefTableFields,
+            FunctionFields,
+            NumKinds
+        };
+
+        Kind kind;
+        unsigned offset : 24;
+
+        inline AbstractHeap(Kind kind_in)
+            : kind(kind_in)
+            , offset(0)
+        { }
+
+        inline AbstractHeap(Kind kind_in, unsigned offset_in)
+            : kind(kind_in)
+            , offset(offset_in)
+        {
+            // Currently offsets are only used for things like globals or
+            // tables, which are bounded by the limits specified in
+            // WasmLimits.h. If we want to expand abstract heaps to some kind
+            // of integer range analysis of pointers into WASM memory we might
+            // want to remove this and unpack the offset.
+
+            RELEASE_ASSERT(offset == offset_in);
+        }
+    };
+
+    template<typename T>
+    struct HeapAndValue {
+        AbstractHeap heap;
+        T* value;
+    };
+
+    MemoryValue* recordHeap(MemoryValue* value, AbstractHeap::Kind kind, unsigned offset = 0)
+    {
+        m_heapMemoryValues.append({ AbstractHeap(kind, offset), value });
+        return value;
+    }
+
+    void computeHeapRanges()
+    {
+        // FIXME: We should consider unifying this with the FTL heap range computation at some point.
+        // https://bugs.webkit.org/show_bug.cgi?id=276505
+
+        std::array<unsigned, AbstractHeap::NumKinds> maxIndicesPerKind;
+
+        // Compute intervals for each heap type. We assume the minimum index for each type is zero.
+        for (unsigned i = 0; i < AbstractHeap::NumKinds; i ++)
+            maxIndicesPerKind[i] = 1; // Each kind should have at least one index.
+        for (auto info : m_heapMemoryValues)
+            maxIndicesPerKind[info.heap.kind] = std::max<unsigned>(maxIndicesPerKind[info.heap.kind], info.heap.offset + 1);
+
+        std::array<unsigned, AbstractHeap::NumKinds> startIndexPerKind;
+        startIndexPerKind[0] = 0;
+        for (unsigned i = 0; i < AbstractHeap::NumKinds - 1; i ++)
+            startIndexPerKind[i + 1] = startIndexPerKind[i] + maxIndicesPerKind[i];
```

Comment:
I believe you can combine this scan op into the prior loop; I don't know how big the array is so it might not matter, but I don't think the compiler will fuse them on its own.

Reply:
Refactored this function in the latest push, so this is no longer relevant.